### ISSUE TITLE
Highlight enclosing indentation?

### ISSUE THREAD

**Issue by gekoke** [jdtsmith/indent-bars#36]
Is there any way to highlight the currently *enclosing* indentation (i.e., one level of indentation above the current one)?

This is the current behavior with `indent-bars-highlight-current-depth` set to `t`:

![image](https://github.com/jdtsmith/indent-bars/assets/63151928/27dbdb29-9871-4292-a7f8-19fafb0080a1)
In this case, I'd like the first bar to be highlighted - not the second.

![image](https://github.com/jdtsmith/indent-bars/assets/63151928/44ad207e-5310-4983-a7a8-61455664afab)
In this case, I'd like the second bar to be highlighted - not the third. In fact, I'd prefer for the last bar to not be render at all.

I understand that `indent-bars-highlight-current-depth` is named quite literally. I would find it much more useful if there was a `indent-bars-highlight-enclosing-depth` provided or similar. I was quite surprised when I realized that that's *not* what the former variable does, since this seems to be standard behavior in other editing environments.

**Comment by gekoke** [comment]
I suppose this is two separate issues: the ability to highlight enclosing space, and the ability to disable rendering of stipples for the last (non-enclosing) level of indentation.

**Comment by jdtsmith** [comment]
Sorry, you want to configure an arbitrary (likely negative) offset from the current indentation depth to _actually_ highlight? The last bar won't be shown if the lines are filled.  This uses a line-by line-algorithm. If your innermost line were a `for ...:` you _would_ want them?

**Comment by gekoke** [comment]
Hi, thanks for your reply!

> Have you tried disabling `display-on-blank-lines`?

I tried toggling the variable, but it doesn't seem like it's relevant (I will set it to `nil` for the sake of clarity henceforth, though).

I think what I mean would best be explained by the following example:

Given the following configuration:
```elisp
(use-package indent-bars ;; at 230e3c8
  :hook
  (prog-mode . indent-bars-mode)
  (conf-mode . indent-bars-mode)
  (json-ts-mode . indent-bars-mode)
  :custom
  (indent-bars-display-on-blank-lines nil)
  (indent-bars-color '(highlight :face-bg t :blend 0.3))
  (indent-bars-pattern ".")
  (indent-bars-width-frac 0.1)
  (indent-bars-pad-frac 0.1)
  (indent-bars-highlight-current-depth '(:blend 1 :width 0.3)))
```

And given this file:
```py
class Foo:
    @property
    def bars(self) -> Iterable[Bar]:
        entries = FooBar.objects.filter(foo=self)
        bars = [Bar(entry.bar) for entry in entries]
        return bars

    @property
    def bar_labels(self) -> Iterable[str]:
        return [bar.label for bar in self.bars]

    def add_bar(self, baz: Bar, quantity: int = 1):
        for __ in range(quantity):
            entry = FooBar()
            entry.foo = self
            entry.bar = baz
            entry.save()
```

indent-bars highlights the following bar:
![image](https://github.com/jdtsmith/indent-bars/assets/63151928/3e77bf67-a00d-42c7-9700-be14aefe559f)

And VSCode highlights the following bar, with the cursor at the same point (note the slightly brighter bar at line 10):
![image](https://github.com/jdtsmith/indent-bars/assets/63151928/20ccb062-c8dd-4dc9-bf84-a322e565f733)

The former is the behavior I would like and that I consider unsurprising.

**Comment by jdtsmith** [comment]
> I tried toggling the variable, but it doesn't seem like it's relevant (I will set it to nil for the sake of clarity henceforth, though).

Did you also `M-x indent-bars-reset` after you altered it that variable?  It's relevant to your statement `I'd prefer for the last bar to not be render at all.`

> The latter is the behavior I would like and that I consider unsurprising.

What happens if you move that VSCode cursor _forward_ between its current position, up to the `return` statement? `indent-bars` bases its "current highlight" on the _indent_ of the current line, not the cursor position (which sounds quite distracting to me if that's what VSCode really does, since it will change as you move around the line).

**Comment by gekoke** [comment]
> Did you also `M-x indent-bars-reset` after you altered it that variable? It's relevant to your statement `I'd prefer for the last bar to not be render at all.`

Yes, I did. I honestly had forgotten that I even mentioned it. That does disable rendering of the last bar, so we're all clear on that.

> What happens if you move that VSCode cursor forward between its current position, up to the return statement? indent-bars bases its "current highlight" on the indent of the current line, not the cursor position (which sounds quite distracting to me if that's what VSCode really does, since it will change as you move around the line).

I didn't manage to capture the cursor on the screenshot, but it's actually at (just before) the `return` keyword on line 10, in the same position as it is in the screenshot with Emacs.

I've attached a video of VSCode's behavior.

https://github.com/jdtsmith/indent-bars/assets/63151928/ead26907-3e29-4f80-ba17-361f04a85dfb

Thanks for your time!

**Comment by jdtsmith** [comment]
Thanks, but it's very hard to follow the video or appreciate the differences.   This is _not_ just "enclosing scope".  For example, when cursor is on the `for __ in` line, the highlighted bar is the one below it (just as in indent-bars).  

I think your request is related to #24: constraining highlighting to stay within the _local scope_.  This is technically challenging given the highlight mechanism used.

I do actually have some capability for that coming soon, so please check again when that becomes available (assuming you use treesitter and are interested).

**Comment by gekoke** [comment]
> Thanks, but it's very hard to follow the video or appreciate the differences. This is _not_ just "enclosing scope". For example, when cursor is on the `for __ in` line, the highlighted bar is the one below it (just as in indent-bars).
> 
> I think your request is related to #24: constraining highlighting to stay within the _local scope_. This is technically challenging given the highlight mechanism used.

Yes, I saw this issue. I specifically am *not* referring to this, though I understand that might not be clear since VSCode *does* support this.

I'm referring to the discrepancy in the bar chosen to highlight in the following situation - note that the cursor is at the same point in both images:

> indent-bars highlights the following bar: ![image](https://private-user-images.githubusercontent.com/63151928/307422775-3e77bf67-a00d-42c7-9700-be14aefe559f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg3OTI3OTgsIm5iZiI6MTcwODc5MjQ5OCwicGF0aCI6Ii82MzE1MTkyOC8zMDc0MjI3NzUtM2U3N2JmNjctYTAwZC00MmM3LTk3MDAtYmUxNGFlZmU1NTlmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI0VDE2MzQ1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTRiNjMwZjE2NmMxYzJmNDhlNjU1MGQ5NTA4ZTlmNTBkNzIyMDQ2MjVjM2Q4OTU3YTQ1Y2U1NWQxNzZmYzdjNTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.SdpKwXqXBmKMBj2VnFZUWRZcuMXi-eWt-BNCa-Z-nCo)
> 
> And VSCode highlights the following bar, with the cursor at the same point (note the slightly brighter bar at line 10): ![image](https://private-user-images.githubusercontent.com/63151928/307422840-20ccb062-c8dd-4dc9-bf84-a322e565f733.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg3OTI3OTgsIm5iZiI6MTcwODc5MjQ5OCwicGF0aCI6Ii82MzE1MTkyOC8zMDc0MjI4NDAtMjBjY2IwNjItYzhkZC00ZGM5LWJmODQtYTMyMmU1NjVmNzMzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI0VDE2MzQ1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY5Y2FhZjg0ZDkwYjE2ODUzM2ZkOGMwMDI4NDRkMmIwOTBkNWU2NTc4MGIxNGQ1YjdlNTFjMjAwNzE1ZWMyY2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.5DXKwktjfV7LVmK6wGE3h2aqafoVD5p5WKTBWGvi6rM)

Essentially, I'd expect the bar under `def` to be highlighted (the same behaviour as VSCode), not the one under the `for`.

Apologies, for the back and forth - I hope I'm not coming across as demanding, just having trouble finding the words to express what I mean clearly.

**Comment by jdtsmith** [comment]
If I added an option to highlight "one bar up" from what is current done, you would not like that either.  Please read carefully my statement about the `for __ in` line; that proves quite clearly that multi-line syntax (and _not just line-by-line indentation_) is being used by VSCode to decide which bar to highlight.   On the `for __ in` line, "one bar up" semantics would highlight the bar under `def`, again differing from VSCode.

I think that you are so used to VSCode syntax-aware highlighting, you don't even realize that that's what it is doing.  It seems "natural" to constrain highlighting to the surrounding syntactic scope (block, function, etc.), but doing so absolutely does require the type of thing #24 is asking for.

**Comment by gekoke** [comment]
> On the `for __ in` line, "one bar up" semantics would highlight the bar under `def`, again differing from VSCode.

I would find that acceptable - I don't expect the package to mirror VSCode's functionality, considering it performs some kind of syntactical analysis. I just think it's more natural for the highlighted indentation to not actually be the current level of indentation.

> If you want to try simple "one bar up" semantics, you could evaluate the following quick advice in `*scratch*` (though beware, this is an internal function which could change):
> 
> ```emacs-lisp
> (advice-add 'indent-bars--current-indentation-depth :filter-return
>             (lambda (d) (max 0 (1- d))))
> ```

Thank you for providing this snippet. I would have tried this myself already were I more comfortable with Elisp.

> My guess is you will find (different) things about this you don't like.

Most likely - I don't really have an intuition for this problem like you have.

**Comment by gekoke** [comment]
With the limited testing I have done, I *much* prefer the highlighting of the indentation level just above the current one - I think it's far more natural and unsurprising.

Yes, the tree-sitter integration does sound promising.

**Comment by jdtsmith** [comment]
Interesting.  I will probably make that an option at some point.  Might even be a better default for TS (since with it, highlight is "contained" by the surrounding scope).

**Comment by jdtsmith** [comment]
The newly released v0.5 now includes tree-sitter based "scope" support, with full styling of the in-scope and out-of-scope bar regions.  In addition, current highlight depth offers 3 methods, including the new default (`context`) which works well for me.  Give a try and let me know of any issues.
